spec/std/openssl/ssl/socket_spec.cr

```diff
@@ -3,36 +3,53 @@ require "socket"
 require "../../spec_helper"
 require "../../../support/ssl"
 
-describe OpenSSL::SSL::Socket do
-  it "returns the cipher that is currently in use" do
-    tcp_server = TCPServer.new(0)
-    server_context, client_context = ssl_context_pair
+private alias Server = OpenSSL::SSL::Socket::Server
+private alias Client = OpenSSL::SSL::Socket::Client
 
-    OpenSSL::SSL::Server.open(tcp_server, server_context) do |server|
-      spawn do
-        OpenSSL::SSL::Socket::Client.open(TCPSocket.new(tcp_server.local_address.address, tcp_server.local_address.port), client_context, hostname: "example.com") do |socket|
-        end
-      end
+private def socket_test(server_tests, client_tests)
```

Comment:
Maybe add empty procs as default values in order to avoid clutter in the samples?

Reply:
> I'd like to have a positive spec on `signature_algorithm'.
> Maybe this could help: https://github.com/sfackler/rust-openssl/blob/487963d17a71dd7f4b239d7284c623f1c6a0b64b/openssl/src/x509/tests.rs#L303

Yeah, but we'd also have to add a way to load a cert from a pem which it looks like they have (`X509::from_pem`) . That's probably doable, but I'll have to figure out which functions do that.

Reply:
Oh right, that's missing. So go ahead if you like, either in this PR or a followup. But it doesn't need to be in here.

Reply:
It might be pretty straightforward, actually. See https://zakird.com/2013/10/13/certificate-parsing-with-openssl for an example. It's probably better to use `PEM_read_bio_X509` though, and we already have `OpenSSL::BIO` for bridging to `IO`.

Reply:
:/ I'm not understanding how to use `BIO`. I tried this

```crystal
# LibCrypto
fun pem_read_bio_x509 = PEM_read_bio_X509(bp : Bio*, x590 : X509*, cb : Void*, u : Void*) : X509
```

```crystal
# Certificate
    def initialize(pem : String)
      bio = OpenSSL::BIO.new(IO::Memory.new(pem))
      # also tried this way in case it mattered if the data was already in the io
      # io = IO::Memory.new(pem.size)
      # bio = OpenSSL::BIO.new io
      # bio.io << pem
      @cert = LibCrypto.pem_read_bio_x509(bio, nil, nil, nil)
      raise "Unable to parse certificate" if @cert.null?
    end
```

```
Certificate.new("-----BEGIN CERTIFICATE-----
MIICsDCC...blah
")
```

but `PEM_read_bio_X509` is always returning null pointers. I looked into using exactly what the link you sent did, but since you have to later `free` the bio, that seems not great

Reply:
If you raise `OpenSSL::Error.new` it shows you the OpenSSL error message: `error:20068079:BIO routines:BIO_gets:unsupported method`.  Seems like the issue is that `BIO_meth_set_gets` is not implemented in `Bio`.
That should make it work:
```diff
--- i/src/openssl/bio.cr
+++ w/src/openssl/bio.cr
@@ -48,6 +48,19 @@ struct OpenSSL::BIO
       1
     end

+    gets = LibCrypto::BioMethodGets.new do |bio, buffer, size|
+      io = Box(IO).unbox(BIO.get_data(bio))
+      io.flush
+      string = io.gets(size)
+      if string
+        p string
+        string.to_slice.copy_to(buffer, string.bytesize)
+        string.bytesize
+      else
+        0
+      end
+    end
+
     ctrl = LibCrypto::BioMethodCtrl.new do |bio, cmd, num, ptr|
       io = Box(IO).unbox(BIO.get_data(bio))

@@ -92,6 +105,7 @@ struct OpenSSL::BIO
         LibCrypto.BIO_meth_set_write(biom, bwrite)
         LibCrypto.BIO_meth_set_read(biom, bread)
       {% end %}
+      LibCrypto.BIO_meth_set_gets(biom, gets)

       LibCrypto.BIO_meth_set_ctrl(biom, ctrl)
       LibCrypto.BIO_meth_set_create(biom, create)
@@ -103,6 +117,7 @@ struct OpenSSL::BIO
       biom.value.name = "Crystal BIO"
       biom.value.bwrite = bwrite
       biom.value.bread = bread
+      biom.value.bgets = gets
       biom.value.ctrl = ctrl
       biom.value.create = create
       biom.value.destroy = destroy
```

But it still runs into an error: `error:0906D064:PEM routines:PEM_read_bio:bad base64 decode` 
Dunno what this is about. `openssl x509 -noout -in cert.pem -text` works perfectly.

Reply:
Yeah it's strange. It seems to be getting the data, since if you screw up the end cert line, it complains about that instead of the base64. Hmmmm

Reply:
This much sounds like a new PR though.